vrx_ign/src/ScoringPlugin.cc

```diff
@@ -0,0 +1,555 @@
+#include <ignition/plugin/Register.hh>
+
+#include "ScoringPlugin.hh"
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+void cb(const ignition::msgs::Contacts &_contacts){return;}
+
+ScoringPlugin::ScoringPlugin()
+    //: gzNode(new std::unique_ptr<ignition::transport::Node()>)
+{
+}
+
+void ScoringPlugin::Configure(const Entity &_entity,
+                           const std::shared_ptr<const sdf::Element> &_sdf,
+                           EntityComponentManager &_ecm,
+                           EventManager &_eventMgr)
+{
+    // this->world = _world
+    this->world = World(_entity); //TODO: verify: Is this how world is accessed?
+    this->sdf = _sdf;
+
+    // SDF.
+    if (!this->ParseSDFParameters())
+    {
+        ignerr << "Scoring disabled" << std::endl;
+        return;
+    }
+    // Initialize spherical coordinates
+    // auto gzSC = _ecm.Component<components::SphericalCoordinates>(this->world)
+    auto gzSC = this->world.SphericalCoordinates(_ecm);
+
+    this->sc.SetLatitudeReference(gzSC->LatitudeReference());
+    this->sc.SetLongitudeReference(gzSC->LongitudeReference());
+    this->sc.SetElevationReference(gzSC->ElevationReference());
+    this->sc.SetHeadingOffset(gzSC->HeadingOffset());
+
+    this->readyTime = std::chrono::duration<double>(this->initialStateDuration);
+    this->runningTime = this->readyTime +  std::chrono::duration<double>(this->readyStateDuration);
+    this->finishTime = this->runningTime + std::chrono::duration<double>(this->runningStateDuration);
+
+    // Prepopulate the task msg.
+    taskMsgName.set_type(ignition::msgs::Any_ValueType::Any_ValueType_STRING);
+    taskMsgState.set_type(ignition::msgs::Any_ValueType::Any_ValueType_STRING);
+    taskMsgReadyTime.set_type(ignition::msgs::Any_ValueType::Any_ValueType_TIME);
+    taskMsgRunningTime.set_type(ignition::msgs::Any_ValueType::Any_ValueType_TIME);
+    taskMsgElapsedTime.set_type(ignition::msgs::Any_ValueType::Any_ValueType_TIME);
+    taskMsgRemainingTime.set_type(ignition::msgs::Any_ValueType::Any_ValueType_TIME);
+    taskMsgTimedOut.set_type(ignition::msgs::Any_ValueType::Any_ValueType_BOOLEAN);
+    taskMsgNumCollisions.set_type(ignition::msgs::Any_ValueType::Any_ValueType_INT32);
+    taskMsgScore.set_type(ignition::msgs::Any_ValueType::Any_ValueType_DOUBLE);
+
+    taskMsgName.set_string_value(this->taskName);
+
+    ignition::msgs::Time readyTimeMsg, runningTimeMsg;
+    
+    auto ready_sec = std::chrono::duration_cast<std::chrono::seconds>(this->readyTime);
+    auto ready_nsec = std::chrono::duration_cast<std::chrono::nanoseconds>(this->readyTime - ready_sec);
+    readyTimeMsg.set_sec(ready_sec.count());
+    readyTimeMsg.set_nsec(ready_nsec.count());
+    this->taskMsgReadyTime.set_allocated_time_value(&readyTimeMsg);
+
+    runningTimeMsg.set_sec(this->runningTime.count());
+
+    auto running_sec = std::chrono::duration_cast<std::chrono::seconds>(this->runningTime);
```

Comment:
I recommend to create a helper function to set the time, as it appears many times.